WIP: Adding Mythic Run Overview to Event Overview Page

DendyA opened the pull request:
The changes included in this PR will show info on all team's best keys for each mythic dungeon on the event overview page. Furthermore, a couple of the functions that were to be duplicated were instead made into utility functions. These are in ```/lib/mythics.ts```

There are still two changes that I need to make. 

---

The first is to fix the following error that shows up in the console:

> Can't perform a React state update on a component that hasn't mounted yet. This indicates that you have a side-effect in your render function that asynchronously later calls tries to update the component. Move this work to useEffect instead.

This occurs when the user navigates to any specific team's overview page. I suspect the issue occurs because of the moving of the clientLoader code to a utility function so that it could be used in the Event Overview page as well. I don't know React / React Remix enough to understand why moving this code causes the error. And why this does not occur when navigating to the event overview page. The data still loads fine.

---

The second change is in the teams map lambda when creating each team's player table, in ```app/routes/event/show/route.tsx```. 

The parsedMythicData, which is an array of each team's mythic data, is currently indexed by using the index of the map function. This was done during testing since that was the easiest way to make sure what I was doing worked. Now I would like to instead combine the team object with the mythic data so that the teams array is still enumerated but the mythic data is correctly included and the array doesn't need to be indexed.

---

A further addition will be to also include the overall team's score, best single dungeon etc. to the Event Overview page as well.

DendyA committed:
Adding myself as an admin to be able to create a test event

DendyA committed:
Updated the team specific pretty display of completed dungeons

DendyA committed:
Added each team's mythic best run dungeon overview to the event overview page.

Additionally, pulled out some functions that were to be copied and made them
util functions instead.

plohkoon commented on app/lib/mythics.ts:
```diff
@@ -0,0 +1,138 @@
+export async function ParseMythicDataPerTeam(players: any, playersPromises: any) {
```
Any is sad

plohkoon replied:
Also stylistically this isn't a component so it should theoretically be `parseMythicDataPerTeam`

DendyA replied:
Agreed, any is very sad. Therefore, every any typing from mythics.ts (and files that use mythics.ts) has been replaced with an actual type.

DendyA replied:
The function name has been renamed correctly

plohkoon commented on app/lib/mythics.ts:
```diff
@@ -0,0 +1,138 @@
+        }
+    }, 0);
+    return [bestSingleScore, mostUnderTime] as const;
+}
```
Should configure your editor to append new lines to save

DendyA replied:
My editor has been updated to append new lines on save.

plohkoon commented on app/lib/session.server.ts:
```diff
@@ -6,7 +6,7 @@ import {
 import { db, Prisma, User } from "./db.server";
 
-const ADMINS = ["plohkoon", "captainfrogs"];
+const ADMINS = ["plohkoon", "captainfrogs", "pointysalad"];
```
lol

DendyA replied:
Was hoping to sneak that in! Nothing to see here! lol

plohkoon commented:
I like your idea of summary on the overview page. I was actually toying with an idea in my head that the current index page for a route should actually be a subpage, like `/event/slug/details`or something and the index page should actually show like a leaderboard and all the teams scores and stuff. Maybe a really quick summary of which players are on the team and their role icon or something.

I was noticing that error about can't update during transition and I thought it was when navigating off of certain pages. It might have something to do with overwriting the promise that the server loader replies with.

plohkoon commented on app/routes/event/show/route.tsx:
```diff
@@ -47,12 +39,40 @@ export async function loader({ request, params: { slug } }: Route.LoaderArgs) {
 
   const teams = organizeTeams(event.teams);
+  const parsedMythicDataArray: any[] = [];
 
-  return { event, teams, isAdmin };
+  const client = await RaiderIOClient.getInstance();
+
+  for(const team of teams) {
+    const playersPromises = team.players.map((player) =>
+        player.playerServer && player.playerName
+            ? client.character.getCharacterProfile({
+              region: "us",
+              realm: player.playerServer,
+              name: player.playerName,
+              fields: {
+                gear: true,
+                mythic_plus_best_runs: true,
+                mythic_plus_alternate_runs: true,
+                mythic_plus_highest_level_runs: true,
+                mythic_plus_recent_runs: true,
+                mythic_plus_previous_weekly_highest_level_runs: true,
+                mythic_plus_weekly_highest_level_runs: true,
+              },
+            })
+            : Promise.resolve(null)
+    );
+
+    const mythicsParsedPerTeam = await ParseMythicDataPerTeam(team.players, playersPromises);
+
+    parsedMythicDataArray.push(mythicsParsedPerTeam.mythicsPromise);
+  }
```
This is potentially getting really weird with the data loading. I'll have to look on my machine but I'm suspicious of what's a promise here and what's not. You may need to wrap the final `parsedMythicDataArray` with a `Promise.all` or `Promise.allSettled` before returning it to the client.

DendyA replied:
The ```team.players.map``` portion of code this comment is referring to has been moved into ```getPlayersPromises()``` in ```mythic.ts```. The return type of the function is an array of promises of  a team's player's data. This can be interspersed with promises of null due to the else clause in the ternary operator.

For example, due to the event's teams having 5 players on them, there will be an array of 5 promises with each being either null or the player's data.

This array of promises is used in ```parseMythicDataPerTeam()``` in the same ```mythics.ts``` file. In this function, it is wrapped in a ```Promise.allSettled```. Essentially it is collapsed into one promise. This function is run in the clientLoader while the ```getPlayersPromises()``` is run in the loader (server side). The reason being that ```getPlayersPromises()``` needs the RaiderIO client, which has to be called server side. The ```parseMythicDataPerTeam()``` function needs to be used in the clientLoader to resolve the early React state update error called out in this PR. The ```parseMythicDataPerTeam()``` function returns a Promise of an array of mythic data or a Promise of null.

DendyA replied:
This will be marked as resolved as the data loads correctly with only one error in the browser's console which is not stemming from this change. More on that in another comment. Marking this as resolved.

DendyA committed:
Made changes to resolve the majority of the issues raised in the PR.

The following are a list of changes included in this commit:
* Removed all uses of any in mythics.ts. Removed other uses of any with respect to this file and its functions.
* Fixed the early React state update before mounting error by moving the resolution of the player's mythic data
  into the clientLoader.
* Other cleanup and formatting changes.

DendyA commented:
~~One larger change that I needed to make was to remove the ```CharacterNS``` namespace in the ```characters.ts``` file. Many of the types defined in this file are used in ```mythics.ts``` but due to them being defined in a namespace, my browser's console was complaining about trying to import a namespace. Essentially, the way the code was structured (i.e. the code in ```mythics.ts``` was defined outside ```characters.ts``` file) caused the mixing of namespace style and module style. According to the documentation, using the modules structure is the preferred method over namespaces now, hence my decision to remove it.~~

See my later comment about undoing this change and not seeing the issue anymore. Therefore this is not relevant anymore and the changes will be reverted in the next commit.

DendyA commented:
The use of the index (detailed in the second point in the PR's description) to extract the relevant mythic data for that team will be a FIXME for a future update. It works fine for now.

plohkoon commented on app/lib/raiderIO/characters.ts:
```diff
@@ -2,346 +2,344 @@ import type { RaiderIOClient } from ".";
 import type { RootNS } from "./index";
 import { fieldsString, HasChildBool } from "./utils";
 
-export namespace CharacterNS {
```
I'm a bit confused why you were having trouble with a NS import. You can also maybe try changing it to `import type {CharacterNS} ...`. The type directive helps sometimes

DendyA replied:
I just undid these changes and with both import type { CharacterNS } and just import { CharacterNS }, it doesn't seem to show the issue in console anymore. I will undo this specific change in my next commit. Marking this as resolved.

plohkoon commented on app/lib/mythics.ts:
```diff
@@ -0,0 +1,218 @@
+    return playersPromises;
+}
+
+export async function parseMythicDataPerTeam(team: DBTeamType | null, playersPromises: (Promise<null> | Promise<CharacterProfilePayload<OptionalCharacterProfilePayloadOptions>>)[]) {
```
I don't think that this will return the correct type information. You probably need to do something with a generic here. I can try to help with this tomorrow.

plohkoon replied:
Presumably the return type gets correctly inferred though?

plohkoon replied:
Could also just replace the generic with the actual type object we're using. In this case:

```Typescript
{
                        gear: true,
                        mythic_plus_best_runs: true,
                        mythic_plus_alternate_runs: true,
                        mythic_plus_highest_level_runs: true,
                        mythic_plus_recent_runs: true,
                        mythic_plus_previous_weekly_highest_level_runs: true,
                        mythic_plus_weekly_highest_level_runs: true,
                    }
```

DendyA replied:
Upon discussion, the return type defined in the function is correct so I will mark this as resolved.

DendyA committed:
Added back in the CharacterNS namespace.

DendyA committed:
Added some more typing and removed some unused imports. Fixed some minor file formatting.

DendyA committed:
Added in team stats to the event overview page, below the specific dungeon information.

DendyA commented:
The error called out in the PR's description does still exist but only when you navigate to a specific player's page. The reason actually stems from devtools and is not related to any of the data the app loads or processes. That's a couple hours of debugging that I would like back, thanks forge42.
